File: ruby/view/mainwindow.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

class MainWindow < NSWindow
  def initialize
    @keyHandler = KeyEventHandler.alloc.init
  end
  
  def sendEvent(e)
    if e.oc_type == NSKeyDown
      return if @keyHandler.processKeyEvent(e) == 1
    end
    super_sendEvent(e)
  end
  
  def setKeyHandlerTarget(target)
    @keyHandler.target = target
  end
  
  def registerKeyHandler_key_modifiers(sel, key, mods)
    @keyHandler.registerSelector_key_modifiers(sel, key, mods)
  end
  
  def registerKeyHandler_character_modifiers(sel, char, mods)
    @keyHandler.registerSelector_character_modifiers(sel, char, mods)
  end
end

File: Classes/Views/MainWindow.m
// LimeChat is copyrighted free software by Satoshi Nakagawa <psychs AT limechat DOT net>.
// You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL.txt).

#import "MainWindow.h"


@implementation MainWindow
{
    KeyEventHandler* _keyHandler;
}

- (id)initWithContentRect:(NSRect)contentRect styleMask:(NSUInteger)windowStyle backing:(NSBackingStoreType)bufferingType defer:(BOOL)deferCreation
{
    self = [super initWithContentRect:contentRect styleMask:windowStyle backing:bufferingType defer:deferCreation];
    if (self) {
        _keyHandler = [KeyEventHandler new];
    }
    return self;
}

- (void)setKeyHandlerTarget:(id)target
{
    [_keyHandler setTarget:target];
}

- (void)registerKeyHandler:(SEL)selector key:(int)code modifiers:(NSUInteger)mods
{
    [_keyHandler registerSelector:selector key:code modifiers:mods];
}

- (void)registerKeyHandler:(SEL)selector character:(UniChar)c modifiers:(NSUInteger)mods
{
    [_keyHandler registerSelector:selector character:c modifiers:mods];
}

- (void)sendEvent:(NSEvent *)e
{
    if ([e type] == NSKeyDown) {
        if ([_keyHandler processKeyEvent:e]) {
            return;
        }
    }

    [super sendEvent:e];
}

@end
